test/algorithm/nuts_unittest.cpp

```diff
@@ -0,0 +1,374 @@
+    EXPECT_EQ(theta[2], theta_prime[2]);
+
+    EXPECT_EQ(output.n, static_cast<size_t>(0));
+    EXPECT_EQ(output.s, true);
+
+    double expected_ham = expected_potential - 0.5 * 2;
+    EXPECT_EQ(output.alpha, std::min(std::exp(expected_ham - ham), 1.));
+    EXPECT_EQ(output.n_alpha, static_cast<size_t>(1));
+}
+
+TEST_F(nuts_build_tree_fixture, build_tree_recursion_plus_no_opt_output)
+{
+    using namespace alg;
+
+    auto ad_expr = ad::constant(-0.5) * 
+                   (ad_vars[0] * ad_vars[0] + 
+                    ad_vars[1] * ad_vars[1] +
+                    ad_vars[2] * ad_vars[2]
+                   ) ;
+
+    ham = 0.; // manually computed current hamiltonian
+
+    auto input = TreeInput(
+        ad_expr, theta, theta_adj, rho, log_u, v,
+        epsilon, ham
+    );
+
+    // custom uniform distribution will always accept candidate
+    // except when optimized for n'' == 0 in the recursion
+    build_tree(input, output, 1, [](const auto&) {return 0;});
+
+    // input theta properly updated
+    EXPECT_DOUBLE_EQ(theta[0], 4.);
+    EXPECT_DOUBLE_EQ(theta[1], 0.);
+    EXPECT_DOUBLE_EQ(theta[2], -4.);
+    EXPECT_DOUBLE_EQ(theta_adj[0], -4.);
+    EXPECT_DOUBLE_EQ(theta_adj[1], 0.);
+    EXPECT_DOUBLE_EQ(theta_adj[2], 4.);
+    EXPECT_DOUBLE_EQ(rho[0], -1.);
+    EXPECT_DOUBLE_EQ(rho[1], 0.);
+    EXPECT_DOUBLE_EQ(rho[2], 1.);
+
+    // check potential (should not have been accepted)
+    double expected_potential = -4;
+    EXPECT_EQ(output.potential, expected_potential);
+    
+    // theta_prime should not have been accepted
+    EXPECT_DOUBLE_EQ(theta_prime[0], -2.);
+    EXPECT_DOUBLE_EQ(theta_prime[1], 0.);
+    EXPECT_DOUBLE_EQ(theta_prime[2], 2.);
+
+    EXPECT_EQ(output.n, static_cast<size_t>(0));
+    EXPECT_EQ(output.s, false);
+
+    double expected_ham = expected_potential - 0.5 * 2;
+    // alpha = alpha' + alpha''
+    EXPECT_EQ(output.alpha, std::min(std::exp(expected_ham - ham), 1.) +
+                            std::min(std::exp(-16 - 0.5 * 2 - ham), 1.));
+    EXPECT_EQ(output.n_alpha, static_cast<size_t>(2));
+}
+
+TEST_F(nuts_build_tree_fixture, find_reasonable_log_epsilon)
+{
+    auto ad_expr = ad::constant(-0.5) * 
+                   (ad_vars[0] * ad_vars[0] + 
+                    ad_vars[1] * ad_vars[1] +
+                    ad_vars[2] * ad_vars[2]
+                   ) ;
+    double eps = alg::find_reasonable_epsilon(ad_expr, theta, theta_adj, 10000);
+    static_cast<void>(eps);
+}
+
+TEST_F(nuts_build_tree_fixture, nuts)
+{
+    constexpr size_t n_samples = 10000;
+    constexpr size_t warmup = 10000;
+    constexpr size_t n_adapt = 1000;
+    double delta = 0.6;
+
+    std::vector<Param<double>> thetas(2);
+
+    std::vector<double> samples_0(n_samples);
+    thetas[0].set_storage(samples_0.data());
+
+    auto model = (
+        thetas[0] |= normal(0., 1.)
+    );
+
+    size_t max_depth = 10;
+    size_t seed = 4821;
+    nuts(model, warmup, n_samples, n_adapt, seed,
```

Comment:
I think we need some more end to end tests like this. Take some of the more complicated regression tests and incorporate them here for NUTS.

Reply:
I know I was tired yesterday. Updated.